### PR TITLE
AST: comment after semicolon

### PR DESCRIPTION
**Pull request by helixbass**
@GeoffreyBooth PR ensuring that a comment after a semicolon gets correctly included in the AST

As of this PR, there are only 2 failing tests in the ESLint plugin (out of 11,000+), neither of which I think are Coffeescript AST issues, so this should conclude reconciling the `ast` branch with the ESLint plugin (ie once this PR gets merged, I'll swap in the `ast` branch as the ESLint plugin's Coffeescript dependency)

Based on `ast-generated-jsx-tokens`, [here](https://github.com/helixbass/copheescript/compare/ast-generated-jsx-tokens...ast-semicolon-comment) is just the diff against that branch

### CODE REVIEW

File: src/rewriter.coffee [hunk 1]
```diff
@@ -685,6 +685,8 @@ exports.Rewriter = class Rewriter
           tokens.splice i, 1, @indentation()...
           return 1
         if @tag(i + 1) in EXPRESSION_CLOSE
+          if token[1] is ';' and @tag(i + 1) is 'OUTDENT'
```
Comment (helixbass) [inline src/rewriter.coffee]: By storing a reference to the `;` token preceding an `OUTDENT`, we can preserve both its comments and location data

File: src/helpers.coffee [hunk 2]
```diff
@@ -117,8 +117,8 @@ buildLocationData = (first, last) ->
 # Build a list of all comments attached to tokens.
 exports.extractAllCommentTokens = (tokens) ->
   allCommentsObj = {}
-  for token in tokens when token.comments
-    for comment in token.comments
+  for token in tokens when token.comments or token.prevToken?.comments
+    for comment in [(token.prevToken?.comments ? [])..., (token.comments ? [])...]
```
Comment (GeoffreyBooth) [inline src/helpers.coffee]: Why is this change necessary? It feels like a change that may have gotten the particular test to pass, but in a larger codebase it might introduce bugs.

In particular, there’s logic in the comments output code that avoids outputting a comment if that comment has already been output. That logic is there because sometimes the same comment gets attached to multiple tokens, such as when tokens are duplicated in the rewriter. So what I’m concerned about is that adding a condition here for looking for comments in not just the current token but also in the previous one too makes _every_ case of comments attached to tokens this “duplicated comments” case, that I’m working around by checking that we don’t double-output comments; but that’s really suboptimal because ideally comments wouldn’t be duplicated in the first place. Like if you removed the “don’t double-output comments” code, I bet that every comment would be double-output now, not just the ones where we expect that problem because of rewriting.

In other words, perhaps something needs to be done in the rewriter to move/copy the comments to/from the semicolon token as appropriate, rather than this global change looking for comments beyond where we “should” be looking.

Comment (helixbass) [reply src/helpers.coffee]: Ok I updated to avoid the need for this change by instead using `moveComments()` to shift the comment (that was getting lost) from the semicolon token to the outdent token

Which then also gets that comment to output during compile-to-JS (which it wasn't previously, eg this comment gets lost on `master`:
```
if a
  ;
  ### comment ###
```

Comment (GeoffreyBooth) [reply src/helpers.coffee]: Good! Thank you, that’s much better.

Doesn’t seem to fix #5241, but it’s on the right track.

**Comment by GeoffreyBooth** [comment]
👍 when upstream PRs clear.
